### PR TITLE
fix: recursive delete: backporting changes from Java

### PR DESCRIPTION
**Pull request by thebrianchen**
Backporting some changes from the Java [port](https://github.com/googleapis/java-firestore/pull/622)

### CODE REVIEW

File: dev/src/recursive-delete.ts [hunk 1]
```diff
@@ -84,6 +84,21 @@ export class RecursiveDelete {
...
   private documentsPending = true;
 
+  /**
+   * Whether run() has been called.
+   * @private
+   */
+  private started = false;
+
+  /** Query limit to use when fetching all descendants. */
+  private maxPendingOps: number;
```
Comment (schmidt-sebastian) [inline dev/src/recursive-delete.ts:94]: Does this need to be marked "private" (here and below)?

File: dev/test/recursive-delete.ts [hunk 2]
```diff
@@ -235,8 +239,30 @@ describe('recursiveDelete() method:', () => {
...
+
+      // This deferred promise is used to delay the BatchWriteResponses from
+      // returning in order to create the situation where the number of pending
+      // operations is less than `minPendingOps`.
+      const bufferDeferred = new Deferred<void>();
+
+      // This deferred completes when the second query is run.
+      const secondQueryDeferred = new Deferred<void>();
+
+      const firstStream = Array.from(Array(maxPendingOps).keys()).map((_, i) =>
```
Comment (schmidt-sebastian) [inline dev/test/recursive-delete.ts:260]: Can you create a helper variable for `Array.from(Array(maxPendingOps).keys())` and name it appropriately? It is used twice and not very pretty :)

**Review by schmidt-sebastian — APPROVED** [review]

File: dev/src/recursive-delete.ts (hunk 1 above)
Comment (thebrianchen) [reply dev/src/recursive-delete.ts:94]: Thanks for catching, done.

File: dev/test/recursive-delete.ts (hunk 2 above)
Comment (thebrianchen) [reply dev/test/recursive-delete.ts]: Done.

**Pull request merged**